### DISCUSSION TITLE
slackclient

### DISCUSSION THREAD

**Discussion by rbignon** [ltworf/localslackirc#386] [General]
Hello,

I've seen you forked the `slackclient` library. Why? What are the differences?

**Comment by ltworf** [comment]
By "forked" you mean almost complete rewrite.

It's not async, it was encoding a bunch of information about the user in the user agent field, it had dependencies. Can't remember what else.

**Reply by rbignon** [reply]
Oh ok, I've just seen the first commit “Forked slackclient”, but yes it seems to be really different now.

I'm working on a big commit, which refactors things, use strings everywhere instead of bytes, and introduces few features and fix few bugs.

**Comment by ltworf** [comment]
Could this all be done in separate commits that are easier to review instead?

**Reply by rbignon** [reply]
It would be difficult for irc.py. Here is the WIP: https://github.com/rbignon/localslackirc/commit/f590fe9c4a30a7307901539ae2a6a822ef978df6

**Reply by rbignon** [reply]
New features:
- display more information at connection
```
Welcome to the Slack Server Teamname, romain!
Your host is domain.slack.com, running version localslackirc-2.0
There are 125 users and 30 bots on 1 server
5 Slack Workspace Admins
105 channels formed
```
- rename user's nickname at connection if not using the right one
```
You're now known as romain
```
- use the slack domain (*.slack.com) instead of `gethostname` for server name and user hostnames
- more information in WHOIS:
```
romain [U03D36HJC14@domain.slack.com]
 ircname  : Romain Bignon
          : Developer
 account  : romain@example.org
 account  : +33123456789
 account  : https://avatars.slack-edge.com/2022-10-27/XXXX_original.png
 channels : #general #random
 server   : domain.slack.com [TeamName]
 away     : My Status
          : Workspace Owner
 idle     : 0 days 23 hours 39 mins 19 secs
End of WHOIS
```
- if I send a message to a user who is away, I get a `RPL_AWAY` message
- add `--thread-replies` parameter to display thread answers in IM/channels instead of creating a new channel
- handle threads in IM
- correctly handle edited/deleted messages in IM
- fix `/me` messages
- `USERHOST` displays away/admin flags
- handle `CAP` command (currently empty)
- handle `JOIN :` command sent by recent irssi versions during connection (I don't know why, but if it doesn't get the right answer it freezes)
- fix crash in some case during shutdown or user disconnection
- handle slack connection errors
- handle `MODE` with parameters to display an error
- handle `MODE +b` to display an emply banlist
- add command `WHOWAS`

Internal changes:
- move *.py files into a new `localslackirc` module (requires to use `python -m localslackirc` to run it locally, it will need to change packaging to create a `localslackirc` script)
- split `irc.py` into `daemon.py` and `irc.py`
- do not use bytes anymore
- rework of the commands handling
- use `logging` module instead of syslog
- do not crash if an exception is uncatched in a command handler, just display an error to user and display the backtrace in logs

**Reply by ltworf** [reply]
Yes that needs to be done in different commits, the diff is impossible to review for me if all is done in the same one commit.

I can give you some kind of access so you can make PR directly on my repo, if that is more convenient

**Reply by rbignon** [reply]
https://github.com/ltworf/localslackirc/pull/387/commits
